File: src/test/java/ru/jerael/booktracker/backend/web/controller/BookControllerTest.java
package ru.jerael.booktracker.backend.web.controller;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.context.annotation.Import;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.assertj.MockMvcTester;
import ru.jerael.booktracker.backend.domain.exception.factory.BookExceptionFactory;
import ru.jerael.booktracker.backend.domain.model.book.*;
import ru.jerael.booktracker.backend.domain.model.image.ImageFile;
import ru.jerael.booktracker.backend.domain.model.pagination.PageQuery;
import ru.jerael.booktracker.backend.domain.model.pagination.PageResult;
import ru.jerael.booktracker.backend.domain.service.token.AuthTokenService;
import ru.jerael.booktracker.backend.domain.usecase.book.*;
import ru.jerael.booktracker.backend.web.config.WebProperties;
import ru.jerael.booktracker.backend.web.dto.book.BookCreationRequest;
import ru.jerael.booktracker.backend.web.dto.book.BookDetailsUpdateRequest;
import ru.jerael.booktracker.backend.web.dto.book.BookResponse;
import ru.jerael.booktracker.backend.web.exception.handler.GlobalExceptionHandler;
import ru.jerael.booktracker.backend.web.mapper.BookWebMapper;
import ru.jerael.booktracker.backend.web.mapper.GenreWebMapper;
import ru.jerael.booktracker.backend.web.mapper.UploadCoverWebMapper;
import ru.jerael.booktracker.backend.web.security.SecurityConfig;
import ru.jerael.booktracker.backend.web.validator.FileValidator;
import tools.jackson.databind.ObjectMapper;
import java.io.ByteArrayInputStream;
import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication;

@WebMvcTest(BookController.class)
@Import({GlobalExceptionHandler.class, BookWebMapper.class, GenreWebMapper.class, SecurityConfig.class})
class BookControllerTest {

    @Autowired
    private MockMvcTester mockMvcTester;

    @MockitoBean
    private WebProperties webProperties;

    @Autowired
    private ObjectMapper objectMapper;

    @MockitoBean
    private GetBooksUseCase getBooksUseCase;

    @MockitoBean
    private GetBookByIdUseCase getBookByIdUseCase;

    @MockitoBean
    private CreateBookUseCase createBookUseCase;

    @MockitoBean
    private UploadCoverUseCase uploadCoverUseCase;

    @MockitoBean
    private DeleteCoverUseCase deleteCoverUseCase;

    @MockitoBean
    private DeleteBookUseCase deleteBookUseCase;

    @MockitoBean
    private UpdateBookDetailsUseCase updateBookDetailsUseCase;

    @MockitoBean
    private GetBookCoverUseCase getBookCoverUseCase;

    @MockitoBean
    private FileValidator fileValidator;

    @MockitoBean
    private BookWebMapper bookWebMapper;

    @MockitoBean
    private UploadCoverWebMapper uploadCoverWebMapper;

    @MockitoBean
    private AuthTokenService authTokenService;

    private final UUID id = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");
    private final UUID userId = UUID.fromString("2c5781ea-1bc2-4561-a83d-26106df2526e");
    private final String title = "title";
    private final String author = "author";
    private final BookStatus status = BookStatus.READING;
    private final Instant createdAt = Instant.ofEpochMilli(1771249699347L);

    private UsernamePasswordAuthenticationToken getAuth() {
        return new UsernamePasswordAuthenticationToken(userId, null, Collections.emptyList());
    }

    @Test
    void getAll_ShouldReturnListOfBookResponses() {
        Book book = new Book(
            id,
            title,
            null,
            createdAt,
            Collections.emptySet(),
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        BookResponse bookResponse = new BookResponse(
            id,
            title,
            null,
            status.getValue(),
            createdAt,
            Collections.emptySet(),
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );

        PageResult<Book> pageResult = new PageResult<>(List.of(book), 10, 0, 1, 1);
        when(getBooksUseCase.execute(any(PageQuery.class), eq(userId))).thenReturn(pageResult);
        when(bookWebMapper.toResponse(book)).thenReturn(bookResponse);

        var mockResponse = mockMvcTester.get().uri("/api/v1/books?page=0&size=10")
            .with(authentication(getAuth())).exchange();

        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .extractingPath("$.content")
            .convertTo(BookResponse[].class)
            .satisfies(responses -> assertThat(responses).containsExactly(bookResponse));

        assertThat(mockResponse)
            .bodyJson()
            .extractingPath("$.page.totalElements").isEqualTo(1);

        verify(getBooksUseCase).execute(any(PageQuery.class), eq(userId));
    }

    @Test
    void getById_WhenExceptionThrown_ShouldReturnNotFound() {
        UUID id = UUID.fromString("31d3f5e3-7faf-4678-a3cf-4657d8875a82");
        when(getBookByIdUseCase.execute(id, userId)).thenThrow(BookExceptionFactory.bookNotFound(id));

        var mockResponse = mockMvcTester.get().uri("/api/v1/books/" + id).with(authentication(getAuth()));

        assertThat(mockResponse)
            .hasStatus(HttpStatus.NOT_FOUND)
            .bodyJson()
            .extractingPath("$.detail")
            .isEqualTo("Book with id " + id + " was not found");

        verify(getBookByIdUseCase).execute(id, userId);
    }

    @Test
    void deleteBook_ShouldReturnNoContent() {
        var mockResponse = mockMvcTester.delete().uri("/api/v1/books/" + id).with(authentication(getAuth()));

        assertThat(mockResponse).hasStatus(HttpStatus.NO_CONTENT);
        verify(deleteBookUseCase).execute(id, userId);
    }

    @Test
    void updateDetails_ShouldReturnUpdatedBook() {
        BookDetailsUpdateRequest request = new BookDetailsUpdateRequest(
            "new title",
            null,
            "want_to_read",
            null
        );
        BookDetailsUpdate data = new BookDetailsUpdate("new title", null, BookStatus.WANT_TO_READ, null);
        Book book = new Book(
            id,
            "new title",
            null,
            createdAt,
            Collections.emptySet(),
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        BookResponse bookResponse = new BookResponse(
            id,
            "new title",
            null,
            "want_to_read",
            createdAt,
            Collections.emptySet(),
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        when(bookWebMapper.toDomain(request)).thenReturn(data);
        when(updateBookDetailsUseCase.execute(id, userId, data)).thenReturn(book);
        when(bookWebMapper.toResponse(book)).thenReturn(bookResponse);

        var mockResponse = mockMvcTester.patch().uri("/api/v1/books/" + id)
            .with(authentication(getAuth()))
            .contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(request));

        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(BookResponse.class)
            .satisfies(response -> {
                assertThat(response.title()).isEqualTo(request.title());
                assertThat(response.status()).isEqualTo(book.status().getValue());
            });
    }

    @Test
    void create_ShouldCreateBook() {
        BookCreationRequest request = new BookCreationRequest(title, author, status.getValue(), Set.of(1, 2));
        BookCreation data = new BookCreation(title, author, null, Set.of(1, 2));
        Book book = new Book(
            id,
            title,
            null,
            createdAt,
            Collections.emptySet(),
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        BookResponse bookResponse = new BookResponse(
            id,
            title,
            null,
            status.getValue(),
            createdAt,
            Collections.emptySet(),
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        when(bookWebMapper.toDomain(request)).thenReturn(data);
        when(createBookUseCase.execute(data, userId)).thenReturn(book);
        when(bookWebMapper.toResponse(book)).thenReturn(bookResponse);

        var mockResponse = mockMvcTester.post().uri("/api/v1/books")
            .with(authentication(getAuth()))
            .contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(request));

        assertThat(mockResponse)
            .hasStatus(HttpStatus.CREATED)
            .bodyJson()
            .convertTo(BookResponse.class)
            .satisfies(response -> {
                    assertThat(response.id()).isEqualTo(bookResponse.id());
                    assertThat(response.title()).isEqualTo(bookResponse.title());
                }
            );
    }

    @Test
    void uploadCover_ShouldUploadCover() {
        String coverFileName = "cover.jpg";
        MockMultipartFile mockMultipartFile = new MockMultipartFile(
            "cover",
            "image.jpg",
            MediaType.IMAGE_JPEG_VALUE,
            "content".getBytes()
        );
        UploadCover data = new UploadCover(MediaType.IMAGE_JPEG_VALUE, null, 0L);
        Book book = new Book(
            id,
            title,
            coverFileName,
            createdAt,
            null,
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        BookResponse bookResponse = new BookResponse(
            id,
            title,
            coverFileName,
            status.getValue(),
            createdAt,
            null,
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        when(uploadCoverWebMapper.toDomain(mockMultipartFile)).thenReturn(data);
        when(uploadCoverUseCase.execute(id, userId, data)).thenReturn(book);
        when(bookWebMapper.toResponse(book)).thenReturn(bookResponse);

        var mockResponse = mockMvcTester.post().uri("/api/v1/books/" + id + "/cover")
            .with(authentication(getAuth()))
            .multipart()
            .file(mockMultipartFile);

        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(BookResponse.class)
            .satisfies(response -> assertThat(response.coverUrl()).isEqualTo(bookResponse.coverUrl()));
    }

    @Test
    void deleteCover_ShouldReturnNoContent() {
        var mockResponse = mockMvcTester.delete().uri("/api/v1/books/" + id + "/cover")
            .with(authentication(getAuth()));

        assertThat(mockResponse).hasStatus(HttpStatus.NO_CONTENT);
        verify(deleteCoverUseCase).execute(id, userId);
    }

    @Test
    void getCover_ShouldReturnImageFile() {
        byte[] content = "content".getBytes();
        String coverFileName = "cover.webp";
        String contentType = "image/webp";
        ImageFile imageFile = new ImageFile(
            coverFileName,
            contentType,
            new ByteArrayInputStream(content),
            content.length
        );
        when(getBookCoverUseCase.execute(id, userId)).thenReturn(imageFile);

        var mockResponse = mockMvcTester.get().uri("/api/v1/books/" + id + "/cover")
            .with(authentication(getAuth()));

        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .hasContentType(MediaType.parseMediaType(contentType))
            .hasHeader(HttpHeaders.CONTENT_DISPOSITION, "filename=\"" + coverFileName + "\"")
            .hasHeader(HttpHeaders.CACHE_CONTROL, "max-age=86400")
            .body().isEqualTo(content);
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/controller/BookControllerTest.kt
package ru.jerael.booktracker.backend.web.controller

import com.ninjasquad.springmockk.MockkBean
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication
import org.springframework.test.web.servlet.assertj.MockMvcTester
import ru.jerael.booktracker.backend.domain.exception.factory.BookExceptionFactory
import ru.jerael.booktracker.backend.domain.model.book.BookStatus
import ru.jerael.booktracker.backend.domain.model.pagination.PageResult
import ru.jerael.booktracker.backend.domain.service.token.AuthTokenService
import ru.jerael.booktracker.backend.domain.usecase.book.*
import ru.jerael.booktracker.backend.factory.auth.AuthWebFactory
import ru.jerael.booktracker.backend.factory.book.BookDomainFactory
import ru.jerael.booktracker.backend.factory.book.BookWebFactory
import ru.jerael.booktracker.backend.factory.file.FileFactory
import ru.jerael.booktracker.backend.factory.image.ImageDomainFactory
import ru.jerael.booktracker.backend.web.config.WebProperties
import ru.jerael.booktracker.backend.web.dto.book.BookResponse
import ru.jerael.booktracker.backend.web.exception.handler.GlobalExceptionHandler
import ru.jerael.booktracker.backend.web.mapper.BookWebMapper
import ru.jerael.booktracker.backend.web.mapper.UploadCoverWebMapper
import ru.jerael.booktracker.backend.web.security.SecurityConfig
import ru.jerael.booktracker.backend.web.validator.FileValidator
import tools.jackson.databind.ObjectMapper
import java.util.*

@WebMvcTest(BookController::class)
@Import(GlobalExceptionHandler::class, SecurityConfig::class)
class BookControllerTest {
    
    @Autowired
    private lateinit var mockMvcTester: MockMvcTester
    
    @MockkBean
    private lateinit var webProperties: WebProperties
    
    @Autowired
    private lateinit var objectMapper: ObjectMapper
    
    @MockkBean
    private lateinit var getBooksUseCase: GetBooksUseCase
    
    @MockkBean
    private lateinit var getBookByIdUseCase: GetBookByIdUseCase
    
    @MockkBean
    private lateinit var createBookUseCase: CreateBookUseCase
    
    @MockkBean
    private lateinit var uploadCoverUseCase: UploadCoverUseCase
    
    @MockkBean
    private lateinit var deleteCoverUseCase: DeleteCoverUseCase
    
    @MockkBean
    private lateinit var deleteBookUseCase: DeleteBookUseCase
    
    @MockkBean
    private lateinit var updateBookDetailsUseCase: UpdateBookDetailsUseCase
    
    @MockkBean
    private lateinit var getBookCoverUseCase: GetBookCoverUseCase
    
    @MockkBean
    private lateinit var fileValidator: FileValidator
    
    @MockkBean
    private lateinit var bookWebMapper: BookWebMapper
    
    @MockkBean
    private lateinit var uploadCoverWebMapper: UploadCoverWebMapper
    
    @MockkBean
    private lateinit var authTokenService: AuthTokenService
    
    private val userId: UUID = UUID.randomUUID()
    private val bookId: UUID = UUID.randomUUID()
    
    @Test
    fun `getAll should return list of BookResponses`() {
        val book = BookDomainFactory.createBook(userId = userId)
        val bookResponse = BookWebFactory.createBookResponse()
        val pageResult = PageResult(listOf(book), 10, 0, 1, 1)
        
        every { getBooksUseCase.execute(any(), eq(userId)) } returns pageResult
        every { bookWebMapper.toResponse(book) } returns bookResponse
        
        val mockResponse = mockMvcTester.get().uri("/api/v1/books?page=0&size=10")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId))).exchange()
        
        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .bodyJson().apply {
                extractingPath("$.content")
                    .convertTo(Array<BookResponse>::class.java)
                    .satisfies({ responses ->
                        assertThat(responses).containsExactly(bookResponse)
                    })
                extractingPath("$.page.totalElements").isEqualTo(1)
            }
        
        verify { getBooksUseCase.execute(any(), eq(userId)) }
    }
    
    @Test
    fun `when Book not found, getById should return 404 Not Found`() {
        every { getBookByIdUseCase.execute(bookId, userId) } throws BookExceptionFactory.bookNotFound(bookId)
        
        val mockResponse = mockMvcTester.get().uri("/api/v1/books/$bookId")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
        
        assertThat(mockResponse)
            .hasStatus(HttpStatus.NOT_FOUND)
            .bodyJson().apply {
                extractingPath("$.detail").isEqualTo("Book with id $bookId was not found")
            }
        
        verify { getBookByIdUseCase.execute(bookId, userId) }
    }
    
    @Test
    fun `deleteBook should return 204 NO CONTENT`() {
        every { deleteBookUseCase.execute(bookId, userId) } just Runs
        
        val mockResponse = mockMvcTester.delete().uri("/api/v1/books/$bookId")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
        
        assertThat(mockResponse).hasStatus(HttpStatus.NO_CONTENT)
        
        verify { deleteBookUseCase.execute(bookId, userId) }
    }
    
    @Test
    fun `updateDetails should return updated Book`() {
        val newTitle = "New Title"
        val newStatus = BookStatus.COMPLETED.value
        
        val request = BookWebFactory.createBookDetailsUpdateRequest(title = newTitle, status = newStatus)
        val data = BookDomainFactory.createBookDetailsUpdate(
            bookId = bookId,
            userId = userId,
            title = newTitle,
            status = BookStatus.fromString(newStatus)
        )
        val book = BookDomainFactory.createBook(id = bookId, userId = userId, title = newTitle)
        val bookResponse = BookWebFactory.createBookResponse(id = bookId, title = newTitle, status = newStatus)
        
        every { bookWebMapper.toDomain(request, bookId, userId) } returns data
        every { updateBookDetailsUseCase.execute(data) } returns book
        every { bookWebMapper.toResponse(book) } returns bookResponse
        
        val mockResponse = mockMvcTester.patch().uri("/api/v1/books/$bookId")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
            .contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(request))
        
        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(BookResponse::class.java)
            .satisfies({ response ->
                assertEquals(newTitle, response.title)
                assertEquals(newStatus, response.status)
            })
        
        verify { updateBookDetailsUseCase.execute(data) }
    }
    
    @Test
    fun `create should return created Book`() {
        val title = " title "
        val cleanedTitle = "title"
        
        val request = BookWebFactory.createBookCreationRequest(title = title)
        val data = BookDomainFactory.createBookCreation(userId = userId, title = title)
        val book = BookDomainFactory.createBook(id = bookId, userId = userId, title = cleanedTitle)
        val bookResponse = BookWebFactory.createBookResponse(id = bookId, title = cleanedTitle)
        
        every { bookWebMapper.toDomain(request, userId) } returns data
        every { createBookUseCase.execute(data) } returns book
        every { bookWebMapper.toResponse(book) } returns bookResponse
        
        val mockResponse = mockMvcTester.post().uri("/api/v1/books")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
            .contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(request))
        
        assertThat(mockResponse)
            .hasStatus(HttpStatus.CREATED)
            .bodyJson()
            .convertTo(BookResponse::class.java)
            .satisfies({ response ->
                assertEquals(bookId, response.id)
                assertEquals(cleanedTitle, response.title)
            })
        
        verify { createBookUseCase.execute(data) }
    }
    
    @Test
    fun `uploadCover should return updated Book`() {
        val fileName = "cover.jpg"
        val coverUrl = "http://localhost:8080/api/v1/books/$bookId/cover"
        
        val file = FileFactory.createMockMultipartFile(originalFileName = fileName)
        val data = BookDomainFactory.createUploadCoverPayload(
            contentType = file.contentType,
            content = file.inputStream,
            size = file.size
        )
        val book = BookDomainFactory.createBook(id = bookId, userId = userId, coverFileName = fileName)
        val bookResponse = BookWebFactory.createBookResponse(id = bookId, coverUrl = coverUrl)
        
        every { fileValidator.validate(file, any()) } just Runs
        every { uploadCoverWebMapper.toDomain(file, bookId, userId) } returns data
        every { uploadCoverUseCase.execute(data) } returns book
        every { bookWebMapper.toResponse(book) } returns bookResponse
        
        val mockResponse = mockMvcTester.post().uri("/api/v1/books/$bookId/cover")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
            .multipart()
            .file(file)
        
        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .bodyJson()
            .convertTo(BookResponse::class.java)
            .satisfies({ response ->
                assertEquals(coverUrl, response.coverUrl)
            })
        
        verify { uploadCoverUseCase.execute(data) }
    }
    
    @Test
    fun `deleteCover should return 204 NO CONTENT`() {
        every { deleteCoverUseCase.execute(bookId, userId) } just Runs
        
        val mockResponse = mockMvcTester.delete().uri("/api/v1/books/$bookId/cover")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
        
        assertThat(mockResponse).hasStatus(HttpStatus.NO_CONTENT)
        
        verify { deleteCoverUseCase.execute(bookId, userId) }
    }
    
    @Test
    fun `getCover should return ImageFile`() {
        val content = "content".toByteArray()
        
        val file = ImageDomainFactory.createImageFile(content = content.inputStream(), size = content.size.toLong())
        
        every { getBookCoverUseCase.execute(bookId, userId) } returns file
        
        val mockResponse = mockMvcTester.get().uri("/api/v1/books/$bookId/cover")
            .with(authentication(AuthWebFactory.createAuthToken(userId = userId)))
        
        assertThat(mockResponse)
            .hasStatus(HttpStatus.OK)
            .hasContentType(MediaType.parseMediaType(file.contentType))
            .hasHeader(HttpHeaders.CONTENT_DISPOSITION, "filename=\"${file.fileName}\"")
            .hasHeader(HttpHeaders.CACHE_CONTROL, "max-age=86400")
            .body().isEqualTo(content)
    }
}